libpysal/graph/_spatial_lag.py

```diff
@@ -1,25 +1,145 @@
-def _lag_spatial(graph, y):
+import numpy as np
+import pandas as pd
+
+
+def _lag_spatial(graph, y, categorical=False, ties='raise'):
     """Spatial lag operator
 
-    If w is row standardized, returns the average of each observation's neighbors;
-    if not, returns the weighted sum of each observation's neighbors.
+    Constructs spatial lag based on neighbor relations of the graph.
+
 
     Parameters
     ----------
     graph : Graph
         libpysal.graph.Graph
     y : array
         numpy array with dimensionality conforming to w
+    categorical : bool
+        True if y is categorical, False if y is continuous.
+    ties : {'raise', 'random', 'tryself'}, optional
+        Policy on how to break ties when a focal unit has multiple
+        modes for a categorical lag.
+        - 'raise': This will raise an exception if ties are
+          encountered to alert the user (Default).
+        - 'random': Will break ties randomly.
+        - 'tryself': Add a self-weight to attempt to break the tie
+          with the focal label. If the self-weight does not break a
+          tie, or the self-weight induces a tie, the tie will be be
+          broken randomly.
+
 
     Returns
     -------
     numpy.array
-        array of numeric values for the spatial lag
+        array of numeric|categorical values for the spatial lag
+
+
+    Examples
+    --------
+    >>> from libpysal.graph._spatial_lag import _lag_spatial
+    >>> import numpy as np
+    >>> from libpysal.weights.util import lat2W
+    >>> from libpysal.graph import Graph
+    >>> graph = Graph.from_W(lat2W(3,3))
+    >>> y = np.arange(9)
+    >>> _lag_spatial(graph, y)
+    array([ 4.,  6.,  6., 10., 16., 14., 10., 18., 12.])
+
+    Row standardization
+    >>> w = lat2W(3,3)
+    >>> w.transform = 'r'
+    >>> graph = Graph.from_W(w)
+    >>> y = np.arange(9)
+    >>> _lag_spatial(graph, y)
+    array([2.        , 2.        , 3.        , 3.33333333, 4.        ,
+           4.66666667, 5.        , 6.        , 6.        ])
+
+
+    Categorical Lag (no ties)
+    >>> y = np.array([*'ababcbcbc'])
+    >>> _lag_spatial(graph, y, categorical=True)
+    array(['b', 'a', 'b', 'c', 'b', 'c', 'b', 'c', 'b'], dtype=object)
+
+    Handling ties
+    >>> y[3] = 'a'
+    >>> np.random.seed(12345)
+    >>> _lag_spatial(graph, y, categorical=True, ties='random')
+    array(['a', 'a', 'b', 'c', 'b', 'c', 'b', 'c', 'b'], dtype=object)
+    >>> _lag_spatial(graph, y, categorical=True, ties='random')
+    array(['b', 'a', 'b', 'c', 'b', 'c', 'b', 'c', 'b'], dtype=object)
+    >>> _lag_spatial(graph, y, categorical=True, ties='tryself')
+    array(['a', 'a', 'b', 'a', 'b', 'c', 'b', 'c', 'b'], dtype=object)
+
     """
     sp = graph.sparse
     if len(y) != sp.shape[0]:
         raise ValueError(
             "The length of `y` needs to match the number of observations "
             f"in Graph. Expected {sp.shape[0]}, got {len(y)}."
         )
+
```

Comment:
```suggestion
   if ( isinstance(y.dtype, CategoricalDtype)
        or pd.api.types.is_object_dtype(y.dtype)
        or pd.api.types.is_bool_dtype(y.dtype)
        or pd.api.types.is_string_dtype(y.dtype)
    ):
        categorical = True
```

This way we can automatically detect categorical data when it is clear. A user would then need to manually specify only if they have int labels for categories.